src/main/java/com/odysseusinc/athena/service/impl/NotificationServiceImpl.java

```diff
@@ -0,0 +1,114 @@
+            Notification current = notificationRepository.findByUserIdAndVocabularyCode(userId, vocabularyCode);
+            if (current == null) {
+                final VocabularyV5 vocabulary = vocabularyRepository.findOne(vocabularyCode);
+                if (vocabulary != null) {
+                    final VocabularyConversion vocabularyConversion = vocabularyConversionRepository.findByIdV5(vocabularyCode);
+                    final Notification newNotification = new Notification(userId, vocabularyConversion, vocabularyCode, vocabulary.getVersion());
+                    notificationRepository.save(newNotification);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void deleteNotificationSubscription(Long userId, String vocabularyCode) {
+        Notification current = notificationRepository.findByUserIdAndVocabularyCode(userId, vocabularyCode);
+        if (current != null) {
+            notificationRepository.delete(current);
+        }
+    }
+
+    /**
+     * this method is required for the transition of already existing subscriptions
+     */
+    @Override
+    public void ensureVocabularyVersionAndCodeAreSet() {
+
+        Map<String, String> vocabularyVersionMap = buildVocabularyVersionMap();
+
+        final List<Notification> existingNotifications = notificationRepository.findByVocabularyCodeIsNullOrActualVersionIsNull();
+
+        for (Notification notification : existingNotifications) {
+            final String vocabularyCode = notification.getVocabularyConversion().getIdV5();
+            notification.setVocabularyCode(vocabularyCode);
+            final String actualVersion = vocabularyVersionMap.get(vocabularyCode);
+            notification.setActualVersion(actualVersion);
+        }
+    }
+
+    @Override
+    public void processUsersVocabularyUpdateSubscriptions(Long userId, Map<String, String> vocabularyVersionMap) {
+
+        final List<Notification> notificationsSubscriptions = notificationRepository.findByUserId(userId);
+
+        List<Notification> changedSubscriptions = new ArrayList<>();
+        for (Notification notification : notificationsSubscriptions) {
+            final String vocabularyCode = notification.getVocabularyCode();
+            if (!equalsIgnoreCase(notification.getActualVersion(), vocabularyVersionMap.get(vocabularyCode))) {
+                changedSubscriptions.add(notification);
+            }
+        }
+
+        if (!changedSubscriptions.isEmpty()) {
+            emailService.sendVocabularyUpdateNotification(userService.get(userId), changedSubscriptions);
+
+            for (Notification notification : changedSubscriptions) {
+                String theLatestVersion = vocabularyVersionMap.get(notification.getVocabularyCode());
+                notification.setActualVersion(theLatestVersion);
+            }
+        }
+
+    }
+
+    @Override
+    public Map<String, String> buildVocabularyVersionMap() {
```

Comment:
Seems this should not be public. Why providing such internal structure is needed?

Reply:
it was attempt to cache this map but I do not like it too...  I have changed to private